hbase-server/src/main/java/org/apache/hadoop/hbase/io/hfile/LruBlockCache.java

```diff
@@ -947,7 +1031,94 @@ public void run() {
+        * So a lot of blocks put into BlockCache but never read,
+        * but spending a lot of CPU resources.
+        * Here we will analyze how many bytes were freed and decide
+        * decide whether the time has come to reduce amount of caching blocks.
+        * It help avoid put too many blocks into BlockCache
+        * when evict() works very active and save CPU for other jobs.
+        * More delails: https://issues.apache.org/jira/browse/HBASE-23887
+        */
+        
+        // First of all we have to control how much time 
+        // has passed since previuos evict() was launched
+        // This is should be almost the same time (+/- 10s)
+        // because we get comparable volumes of freed bytes each time.
+        // 10s because this is default period to run evict() (see above this.wait)
+        long stopTime = System.currentTimeMillis();
+        if (stopTime - startTime <= 1000 * 10 - 1) {
+          mbFreedSum += bytesFreed/1024/1024;
+        } else {
+          // Here we have to calc what situation we have got.
+          // We have the limit "hbase.lru.cache.heavy.eviction.bytes.size.limit"
+          // and can calculte overhead on it.
+          // We will use this information to decide, 
+          // how to change percent of caching blocks.
+          freedDataOverheadPercent =
+            (int) (mbFreedSum * 100 / cache.heavyEvictionMbSizeLimit) - 100;
+          if (freedDataOverheadPercent > 0) {
+            // Now we are in the situation when we are above the limit
+            // But maybe we are going to ignore it because it will end quite soon
+            heavyEvictionCount++;
+            if (heavyEvictionCount > cache.heavyEvictionCountLimit) {
+              // It is going for a long time and we have to reduce of caching
+              // blocks now. So we calculate here how many blocks we want to skip.
+              // It depends on: 
+              // 1. Overhead - if overhead is big we could more aggressive
+              // reducing amount of caching blocks.
+              // 2. How fast we want to get the result. If we know that our 
+              // heavy reading for a long time, we don't want to wait and can 
+              // increase the coefficient and get good performance quite soon.
+              // But if we don't sure we can do it slowly and it could prevent 
+              // premature exit from this mode. So, when the coefficient is 
+              // higher we can get better performance when heavy reading is stable.
+              // But when reading is changing we can adjust to it and set 
+              // the coefficient to lower value.
+              int ch = (int) (freedDataOverheadPercent * cache.heavyEvictionOverheadCoefficient);
+              // But practice shows that 15% of reducing is quite enough.
+              // We are not greedy (it could lead to premature exit).
+              ch = ch > 15 ? 15 : ch;
```

Comment:
Well, whats the use of the coefficient if you are capping the max reduction to 15%? How often did you hit this in your benchmarks?

Reply:
Almost always few first times more than 15%. I think it is enough  because after 30 sec will be not more 55% and this is any way very good for performance. And no risk of overshooting.